src/adapter/src/coord/introspection.rs

```diff
@@ -0,0 +1,173 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Special cases related to the "introspection" of Materialize
+//!
+//! Every Materialize deployment has a pre-installed [`mz_introspection`] cluster, which
+//! has several indexes to speed up common introspection queries. We also have a special
+//! `mz_introspection` role, which can be used by support teams to diagnose a deployment.
+//! For each of these use cases, we have some special restrictions we want to apply. The
+//! logic around these restrictions is defined here.
+//!
+//!
+//! [`mz_introspection`]: https://materialize.com/docs/sql/show-clusters/#mz_introspection-system-cluster
+
+use mz_repr::GlobalId;
+use mz_sql::catalog::SessionCatalog;
+use mz_sql::plan::Plan;
+
+use crate::catalog::{Catalog, Cluster};
+use crate::notice::AdapterNotice;
+use crate::rbac;
+use crate::session::Session;
+
+use crate::{
+    catalog::builtin::{MZ_INTROSPECTION_CLUSTER, MZ_INTROSPECTION_ROLE},
+    AdapterError,
+};
+
+/// Checks whether or not we should automatically run a query on the `mz_introspection`
+/// cluster, as opposed to whatever the current default cluster is.
+pub fn auto_run_on_introspection<'a, 's>(
+    catalog: &'a Catalog,
+    session: &'s mut Session,
+    depends_on: impl IntoIterator<Item = GlobalId>,
+) -> Option<&'a Cluster> {
+    // If this feature is disabled via LaunchDarkly, or the user has disabled it for
+    // this session.
+    if !catalog.system_config().enable_force_introspection_cluster()
+        || !session.vars().force_introspection_cluster()
+    {
+        return None;
+    }
+
+    // Check to make sure our iterator contains atleast one element, this prevents us
+    // from always running empty queries on the mz_introspection cluster.
+    let mut depends_on = depends_on.into_iter().peekable();
+    let non_empty = depends_on.peek().is_some();
+
+    // Make sure we only depend on the system catalog.
+    let system_only = depends_on.all(|id| {
+        let entry = catalog.get_entry(&id);
+        let schema = &entry.name().qualifiers.schema_spec;
+        catalog.state().is_system_schema_specifier(schema)
+    });
+
+    // If we're allowed to run on the mz_introspection cluster, make sure we can resolve it.
```

Comment:
I think this comment is no longer accurate.

Reply:
Good catch, removed it!